### PR TITLE
fix: prevent UPDATE broadcast NoChange false positive

### LINKED ISSUE #2301: fix(update): broadcast decision made after local state commit causes NoChange false positive
## Problem

Contract updates are never broadcast to subscribers or proximity cache neighbors because the "change detection" logic runs **after** the local state has already been committed. This causes the comparison to always return `NoChange`, effectively breaking all update propagation.

### User Impact

- Messages sent in River (chat app) are not visible to other users
- Contract state updates only affect the local node
- Subscription tree and proximity cache mechanisms are bypassed entirely
- Multi-user applications built on Freenet cannot sync state

### Observed Behavior

When a client sends an update:

1. **20:29:40.574** - Local state is updated: `Contract state updated, new_size_bytes: 1308, phase: "update_complete"`
2. **20:29:40.588** - Network operation starts: `Created new operation - starting network request`
3. **20:29:40.622** - Change detection fails: `UPDATE resulted in NoChange`

The ~48ms gap between steps 1 and 3 shows the state is committed before the broadcast decision.

### Expected Behavior

When a client updates contract state:
1. Change detection should compare incoming state vs **pre-update** state
2. If changed, broadcast to subscribers and proximity cache neighbors
3. Then commit the new state locally

Or alternatively: capture the "changed" flag at update time and use it for broadcast decision.

### Code Flow

1. Client sends `UpdateRequest` to local node
2. `ContractExecutor::upsert_contract_state()` is called in `runtime.rs`
3. State is persisted and logs `"Contract state updated"` (line ~928)
4. Network update operation starts via `request_router`
5. `update.rs` calls change detection which compares incoming state to **current** (already-updated) state
6. Result: `NoChange` because they're now identical

### Key Code Locations

- **State commit**: `crates/core/src/contract/executor/runtime.rs` - `upsert_contract_state()` function around line 36
- **Change detection**: `crates/core/src/operations/update.rs` - multiple locations check for change:
  - Line 240-244: `"UPDATE yielded no state change, skipping broadcast"`
  - Line 396-401: `"BroadcastTo update produced no change, ending propagation"`
  - Line 1031-1036: `"Local update resulted in no change; finishing without broadcast"`
- **Broadcast targeting**: `crates/core/src/operations/update.rs` - `get_broadcast_targets_update()` at line 577

### Log Evidence

From technic peer logs:
```
2025-12-15T20:29:40.574743Z INFO freenet::contract::executor::runtime: Contract state updated, contract: HggS9YdhiKSApiwyVAhjftJaq7dTUQmzo2jU2BTta7BC, new_size_bytes: 1308, phase: "update_complete"

2025-12-15T20:29:40.588414Z INFO freenet::node::request_router: Created new operation - starting network request, transaction: 01KCHV8PBCHMQWBY49ZTWNDP04, resource: Update { key: ContractKey { instance: ContractInstanceId("HggS9YdhiKSApiwyVAhjftJaq7dTUQmzo2jU2BTta7BC"), ... }

2025-12-15T20:29:40.622716Z INFO freenet::contract: UPDATE resulted in NoChange, fetching current state to return UpdateResponse, contract: HggS9YdhiKSApiwyVAhjftJaq7dTUQmzo2jU2BTta7BC, phase: "update_no_change"
```

No `UPDATE_PROPAGATION` or `BroadcastTo` messages ever appear in logs.

## Test Environment

- Two peers: technic (136.62.52.28) and Lukas (81.170.132.66)
- Both connected directly (peer-to-peer) and via gateway
- Both subscribed to same contract `HggS9YdhiKSApiwyVAhjftJaq7dTUQmzo2jU2BTta7BC`
- Proximity cache working correctly - both peers announced contract to each other
- Gateway has both peers registered as subscribers
- Freenet version: 0.1.46

### Gateway Also Failing

The gateway was also failing updates with `missing contract` error because it registered subscribers without storing the contract:
```
ERROR freenet::operations::update: Failed to update contract value, contract: HggS9Ydh..., error: missing contract
```

This is a separate issue but compounds the problem - even if peer-to-peer proximity cache worked, the subscription tree via gateway would still fail.

### Option 1: Capture Change Flag at Commit Time

When `upsert_contract_state()` commits the state, return a `changed: bool` flag indicating whether the state actually changed. Pass this flag to the network operation so it knows to broadcast regardless of current state comparison.

### Option 2: Compare Before Commit

Restructure the flow so change detection happens before `upsert_contract_state()` commits. Only commit after broadcast decision is made.

### Option 3: Store Pre-Update State

Before updating, snapshot the current state. Use this snapshot for change detection in the network operation, then discard it.

## Related

- Proximity cache implementation: `crates/core/src/node/proximity_cache.rs`
- Subscription tree: `crates/core/src/ring/seeding.rs`
- The proximity cache correctly tracks which neighbors have contracts, but `get_broadcast_targets_update()` is never reached because the operation exits early due to `NoChange`

[AI-assisted - Claude]

### PR DESCRIPTION
**Pull request by sanity**
## Problem

Contract updates are never broadcast to subscribers because change detection runs AFTER the local state has already been committed, causing the comparison to always return NoChange.

**Flow before this fix:**
1. Client sends UPDATE to peer-a
2. `perform_contract_update` calls `get_updated_state` → commits state locally
3. Network operation starts with `new_state`
4. Network operation's `update_contract` fetches "previous_state"
5. But "previous_state" is already the updated state!
6. Comparison: `new_state == previous_state` → NoChange (false positive)
7. **No broadcast to subscribers** - updates never propagate

This was visible in logs:
```
20:29:40.574 - "Contract state updated" (state committed locally)
20:29:40.588 - "Created new operation" (network op starts)
20:29:40.622 - "UPDATE resulted in NoChange" (false positive!)
```

## Root Cause Analysis (5-Whys)

1. **Why** does update broadcast fail? → `changed: false` is returned, so broadcast skipped
2. **Why** is `changed: false`? → `update_contract` compares incoming state to current state, they're identical
3. **Why** are they identical? → State was already committed before the network operation started
4. **Why** was it committed early? → `perform_contract_update` calls `get_updated_state` which commits via `attempt_state_update`
5. **Why?** → Architectural issue: two competing update paths both try to commit the state

## Solution

Separate "compute state" from "commit state" in network mode by adding a new `compute_state_update()` method:

1. **Added `compute_state_update()`** - Applies contract update logic WITHOUT committing to storage
2. **Modified `perform_contract_update` for network mode:**
   - Uses `compute_state_update()` instead of `get_updated_state()`
   - If no change detected, returns early (no network op needed)
   - If changed, passes computed state to network operation
   - Network operation handles commit + change detection + broadcast correctly
3. **Local mode unchanged** - Still uses `get_updated_state()` for local-only updates

## Additional Fixes

- **Removed redundant `state_store.update()`** in `get_updated_state()` - `attempt_state_update` already commits the state
- **Removed duplicate notification** in `perform_contract_update` - `attempt_state_update` already sends notifications

## Testing

Added regression test: `test_update_broadcast_propagation_issue_2301`
- Sets up gateway + peer nodes
- Peer puts contract, gateway caches it via GET
- Peer sends UPDATE with new state
- **Verifies gateway receives the update via network broadcast**

The test passes with the fix and would fail without it (gateway would have stale state).

## Verification

All existing tests pass:
- `test_update_contract` ✅
- `test_isolated_node_update_operation` ✅
- `test_update_no_change_notification` ✅ (for actual no-change cases)
- `test_update_broadcast_propagation_issue_2301` ✅ (new regression test)

## Fixes

Closes #2301

[AI-assisted - Claude]

### CODE REVIEW

**Comment by sanity** [comment]
@claude Please do a skeptical review of this PR

File: crates/core/tests/operations.rs [hunk 1]
```diff
@@ -2908,3 +2908,253 @@ async fn test_update_no_change_notification(ctx: &mut TestContext) -> TestResult
...
+            Err(_) => {
+                tracing::warn!("Gateway poll: timeout");
+            }
+        }
+
+        tracing::debug!(
+            "Gateway state not yet updated, polling... (elapsed: {:?})",
+            poll_start.elapsed()
+        );
+    }
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/tests/operations.rs:3161]: This polling approach is a reasonable pattern for testing eventually-consistent distributed systems. The explicit timeout with state verification provides clear failure diagnostics. [AI-assisted - Claude]

File: crates/core/tests/operations.rs [hunk 2]
```diff
@@ -2908,3 +2908,253 @@ async fn test_update_no_change_notification(ctx: &mut TestContext) -> TestResult
...
+/// 4. Verifying that gateway's cached state is updated via broadcast
+///
+/// If the NoChange bug is present, gateway would still have the old state
+/// because the UPDATE broadcast would be skipped.
+#[freenet_test(
+    nodes = ["gateway", "peer-a"],
+    timeout_secs = 180,
+    startup_wait_secs = 15,
+    tokio_flavor = "multi_thread",
+    tokio_worker_threads = 4
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/tests/operations.rs:2936]: Addressed - extracted to named constants `POLL_INTERVAL` and `POLL_TIMEOUT` with documentation explaining the rationale for each value. See commit cc26ee9. [AI-assisted - Claude]

File: crates/core/src/contract/executor/runtime.rs [hunk 3]
```diff
@@ -851,26 +865,132 @@ impl Executor<Runtime> {
...
+        };
+
+        let new_state = WrappedState::new(new_state.into_bytes());
+        let changed = new_state.as_ref() != current_state.as_ref();
+
+        if changed {
+            Ok(ComputedStateUpdate::Changed(new_state))
+        } else {
+            Ok(ComputedStateUpdate::NoChange(current_state.clone()))
+        }
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/src/contract/executor/runtime.rs:1073]: Valid point about duplication. The key difference is that `compute_state_update()` deliberately avoids committing to state_store, while `attempt_state_update()` commits immediately. Refactoring to share logic would require a parameter to control commit behavior, which may add complexity for marginal benefit. Open to suggestions if there's a cleaner approach. [AI-assisted - Claude]

File: crates/core/src/contract/executor/runtime.rs [hunk 4]
```diff
@@ -851,26 +865,132 @@ impl Executor<Runtime> {
...
+                    "Update requires related contracts, using fallback path"
+                );
+                let new_state = self
+                    .get_updated_state(&parameters, current_state, key, updates)
+                    .await?;
+                let summary = self
+                    .runtime
+                    .summarize_state(&key, &parameters, &new_state)
+                    .map_err(|e| ExecutorError::execution(e, None))?;
+                // Note: notification is sent by attempt_state_update
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/src/contract/executor/runtime.rs]: Addressed - added comprehensive documentation explaining that the Changed case returns the computed state to the caller (`perform_contract_update`) which then starts the network operation, and the network op's UpdateQuery handler commits the state AND triggers PeerDB notification to interested subscribers. See commit cc26ee9. [AI-assisted - Claude]

File: crates/core/tests/operations.rs [hunk 5]
```diff
@@ -2908,3 +2908,253 @@ async fn test_update_no_change_notification(ctx: &mut TestContext) -> TestResult
...
+    let peer_a = ctx.node("peer-a")?;
+    let gateway = ctx.node("gateway")?;
+    let ws_api_port_peer_a = peer_a.ws_port;
+    let ws_api_port_gateway = gateway.ws_port;
+
+    tracing::info!("Peer A data dir: {:?}", peer_a.temp_dir_path);
+    tracing::info!("Gateway data dir: {:?}", gateway.temp_dir_path);
+
+    // Give extra time for peer to connect to gateway
+    tokio::time::sleep(Duration::from_secs(5)).await;
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/tests/operations.rs]: The connection establishment delay is necessary because the WebSocket connection involves async handshake. This is a one-time cost per test, not in the hot path. Could be improved with a proper connection-ready check but the fixed sleep is adequate for test purposes. [AI-assisted - Claude]

File: crates/core/src/contract/executor/runtime.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/src/contract/executor/runtime.rs:1073]: Addressed - added comment explaining that this byte comparison leverages CRDT idempotency guarantees: for Freenet's CRDT contracts, if merge(current, incoming) produces bytes identical to current, the incoming state contained no new information. This is different from a semantic equality check. See commit cc26ee9. [AI-assisted - Claude]

File: crates/core/src/contract/executor/runtime.rs [hunk 6]
```diff
@@ -851,26 +865,132 @@ impl Executor<Runtime> {
...
+                let request = UpdateContract { key, new_state };
+                let _op: operations::update::UpdateResult = self.op_request(request).await?;
+                Ok(ContractResponse::UpdateResponse { key, summary }.into())
+            }
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/src/contract/executor/runtime.rs]: Acknowledged - added a TODO comment documenting this known limitation. The MissingRelated path still has the original bug because it falls through to `get_updated_state()`. This is acceptable because related contract scenarios are rare, but should be addressed in a follow-up if it becomes an issue. See commit cc26ee9. [AI-assisted - Claude]

File: crates/core/tests/operations.rs [hunk 7]
```diff
@@ -2908,3 +2908,253 @@ async fn test_update_no_change_notification(ctx: &mut TestContext) -> TestResult
...
+    tracing::info!("Step 4: Verifying gateway received the update via polling GET");
+
+    // Poll gateway for up to 30 seconds to see if the state was updated
+    let poll_start = std::time::Instant::now();
+    let poll_timeout = Duration::from_secs(30);
+    let mut gateway_received_update = false;
+
+    while poll_start.elapsed() < poll_timeout {
+        // Small delay between polls
+        tokio::time::sleep(Duration::from_secs(2)).await;
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/tests/operations.rs]: Addressed - extracted to `POLL_INTERVAL` constant with documentation explaining the 100ms value balances responsiveness vs. system load. See commit cc26ee9. [AI-assisted - Claude]

File: crates/core/tests/operations.rs [hunk 8]
```diff
@@ -2908,3 +2908,253 @@ async fn test_update_no_change_notification(ctx: &mut TestContext) -> TestResult
...
+    tracing::info!("Step 3: Peer-a updating contract with new state");
+
+    // Create updated state (add a todo item)
+    let mut todo_list: test_utils::TodoList =
+        serde_json::from_slice(wrapped_initial_state.as_ref()).unwrap_or_else(|_| {
+            test_utils::TodoList {
+                tasks: Vec::new(),
+                version: 0,
+            }
+        });
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/tests/operations.rs]: Addressed - changed to `expect()` for fail-fast behavior. Test failures should surface immediately rather than being silently converted to fallback values. See commit cc26ee9. [AI-assisted - Claude]

File: crates/core/tests/operations.rs [hunk 9]
```diff
@@ -2908,3 +2908,253 @@ async fn test_update_no_change_notification(ctx: &mut TestContext) -> TestResult
...
+
+                    if !state_on_gateway.tasks.is_empty()
+                        && state_on_gateway.tasks[0].title == "Issue 2301 regression test"
+                    {
+                        tracing::info!(
+                            "SUCCESS: Gateway received the UPDATE broadcast! State updated with {} task(s)",
+                            state_on_gateway.tasks.len()
+                        );
+                        gateway_received_update = true;
+                        break;
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/tests/operations.rs:3134]: Addressed - added `tracing::warn!()` to log unexpected state values. This provides visibility into test behavior without asserting on intermediate states that may be valid during convergence. See commit cc26ee9. [AI-assisted - Claude]

File: crates/core/tests/operations.rs [hunk 10]
```diff
@@ -2908,3 +2908,253 @@ async fn test_update_no_change_notification(ctx: &mut TestContext) -> TestResult
...
+            bail!("Peer-a: Error receiving update response: {}", e);
+        }
+        Err(_) => {
+            bail!("Peer-a: Timeout waiting for update response");
+        }
+    }
+
+    // Step 4: Verify gateway received the update by polling GET
+    // The update should have been broadcast to gateway. We poll GET
+    // because subscription notifications across nodes are known to be flaky.
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/tests/operations.rs]: The test is designed to verify the fix for issue #2301 - that updates propagate via broadcast. The polling with timeout approach handles the inherent timing variability in distributed systems. If this proves flaky in CI, we can adjust timeouts or add retry logic, but initial testing shows it's stable. [AI-assisted - Claude]

**Comment by sanity** [comment]
## Response to Review Feedback

Thanks for the thorough review. I've addressed the feedback in commit cc26ee9:

### Code Changes
- **MissingRelated fallback**: Added TODO comment documenting the known limitation (still has the original bug, but related contract scenarios are rare)
- **Notification flow**: Added comprehensive documentation explaining how Changed case flows through network op to PeerDB notification
- **Byte comparison**: Added comment explaining CRDT idempotency semantics

### Test Improvements
- **Named constants**: Extracted `POLL_INTERVAL` (100ms) and `POLL_TIMEOUT` (30s) with documentation
- **Fail-fast**: Changed `unwrap_or_else` to `expect()` for immediate failure surfacing
- **Logging**: Added `tracing::warn!()` for unexpected state values during polling
- **Configuration docs**: Added comments explaining test attribute values

### Acknowledged but Not Changed
- **Code duplication**: The key difference between `compute_state_update()` and `attempt_state_update()` is commit behavior. Refactoring to share logic would require a parameter to control commit, adding complexity for marginal benefit.
- **Fixed sleeps**: Adequate for test purposes; can be improved if flakiness emerges.

[AI-assisted - Claude]

**Comment by sanity** [comment]
## Response to Skeptical PR Review

Thanks for the thorough analysis. Let me address each concern:

### 1. Double Computation in Network Mode

The reviewer correctly identifies that state is computed twice:
1. Once in `compute_state_update` (no commit)
2. Once when `UpdateQuery` handles the request via `upsert_contract_state`

**This is intentional but worth documenting.** The `UpdateQuery` handler uses the passed `new_state` for change detection, then calls `upsert_contract_state` which will re-compute via `attempt_state_update`. 

**Key insight:** For Freenet's CRDT contracts, update functions MUST be deterministic - `merge(current, delta)` always produces the same output for the same inputs. This is a fundamental requirement of CRDTs, not a new constraint introduced by this PR. Non-deterministic updates would already break Freenet's consistency model.

**However:** The reviewer raises a valid point that we're re-computing unnecessarily. A follow-up could pass the computed state directly to `upsert_contract_state` to avoid redundant computation.

### 2. MissingRelated Fallback Still Has Bug

**Acknowledged and documented.** I added a TODO comment in commit cc26ee9. The reviewer is correct that this path still has the original bug.

**Rationale for shipping as-is:**
- Related contract scenarios are rare in current usage
- The fix significantly improves the common case (simple updates)
- Properly fixing MissingRelated requires more architectural work (multiple round-trips to fetch related contracts without committing intermediate state)

If this becomes a blocker, we should create a follow-up issue.

### 3. Double Notification Removed

The old code had `send_update_notification` called unconditionally, but `attempt_state_update` also sends notifications. This was indeed causing duplicate notifications. The removal is correct.

### 4. Test Coverage

The reviewer correctly notes the test doesn't cover:
- MissingRelated path
- Concurrent updates
- Race conditions

**Response:** The test is specifically a regression test for issue #2301 (the common case). Comprehensive integration tests for edge cases should be separate. The polling approach is pragmatic for verifying eventual consistency.

### 5. Architecture Questions

> "Why not fix at the network operation level?"

Good question. Fetching "previous state" before `UpdateQuery` would require plumbing that context through the operation machinery. The current approach (compute-then-pass) was simpler to implement correctly.

> "Thread safety / race conditions"

Valid concern. Between `compute_state_update` and network op completion, another update could arrive. However, Freenet's CRDT merge semantics handle concurrent updates - the final state will be consistent regardless of order. The change detection might occasionally be wrong (seeing NoChange when there was a concurrent change), but this is an existing race that affects all update paths.

### Summary

I recommend:
1. **Merge as-is** - fixes the common case reported in #2301
2. **Create follow-up issue** for MissingRelated fix
3. **Document** CRDT determinism requirement (already implicit in Freenet's design)

The core fix is sound. The edge cases identified are real but don't block the primary fix.

[AI-assisted - Claude]

**Comment by sanity** [comment]
## MissingRelated Case Fixed (Commit 8229ed6)

Addressed the critical concern from the skeptical review. The MissingRelated case now:

1. **Fetches related contracts WITHOUT committing** the main contract's state
2. **Retries `compute_state_update`** with the fetched related contracts in the updates list
3. **Loops until** Changed or NoChange, then proceeds to network operation

This mirrors the loop structure in `get_updated_state` but uses `compute_state_update` (no commit) instead of `attempt_state_update` (commits immediately).

The fix ensures that updates requiring related contracts will also correctly broadcast to peers.

[AI-assisted - Claude]
